security/pkg/pki/ra/common.go

```diff
@@ -86,6 +89,76 @@ func ValidateCSR(csrPEM []byte, subjectIDs []string) bool {
 			return false
 		}
 	}
+
+	// Verify Istio CSR was created by Istio-Proxy by generating template using same the
+	// same process as performed on the client and comparing the generated CSR with the
+	// received CSR.
+
+	// Convert csrIDs to Host comma separated string
+	hosts := strings.Join(csrIDs, ",")
+	// genCSRTemplate is the expected CSR template. The template's extensions are marshaled
+	// in the ExtraExtensions field. CreateCertificateRequest would normally add the SAN extensions
+	// from ExtraExtensions to the Extensions field. However, we are only generating the
+	// template here and not the actual CSR since we do not know the exact signing mechanisms
+	// of the client. Additionally, if we compared extensions the the generated CSR and the orginal
+	// CSR would match since the hosts where constructed from the extracted SAN extensions.
+	genCSRTemplate, err := util.GenCSRTemplate(util.CertOptions{Host: hosts})
+	if err != nil {
+		return false
+	}
+	if !compareCSRs(csr, genCSRTemplate) {
+		return false
+	}
+	return true
+}
+
+// Compare Workload CSRs
+// Verification of the CSR fields based on expected setting from Istio CSR creation in
+// security/pkg/pki/util/generate_csr.go and security/pi/nodeagent/cache/secretcache.go
+func compareCSRs(orgCSR, genCSR *x509.CertificateRequest) bool {
+	// Compare the CSR fields
+	if orgCSR == nil || genCSR == nil {
+		return false
+	}
+	if orgCSR.Subject.CommonName != genCSR.Subject.CommonName {
+		return false
+	}
+	// Acceptable length is O or 1. GenCSR creates a CSR with an Org value of ""
+	if len(orgCSR.Subject.Organization) > 1 {
+		return false
+	}
+	// if the orgCSR has an organization, it should equal ""
+	if len(orgCSR.Subject.Organization) == 1 && orgCSR.Subject.Organization[0] != "" {
+		return false
+	}
+	// Expected length is 0 or 1
+	if len(orgCSR.URIs) > 1 {
+		return false
+	}
+	// Expected length is 0
+	if len(orgCSR.EmailAddresses) != len(genCSR.EmailAddresses) {
```

Comment:
I think there are many dozens of fields ( and possible extensions ) in the X509 spec.

It is better to serialize the subject and compare the result.

Reply:
I ran into testing challenges with this approach. When generating certs requests using `openssl` I was unable to set the `Org` to a string array containing an empty string. I needed to generate cert requests this way to set certain fields for testing (for example, CA in basic constraints and DNS Name). If I compared the serialized subjects this way the tests would always fail. because the orgs wouldn't match. I think I am more concerned about testability here than perfect coverage of all subject fields. What are your thoughts?

Reply:
More concerned about security and testing what we use than made up CSRs we don't generate or use - the agent always generates the same thing, and the DER encoding should be the same, that's what D is supposed to mean...

Reply:
The made up CSRs allow us to test a larger variety of CSR settings in the broader CSR verification workflow.

Reply:
@bleggett Do you have a preference on comparing the serialized subject vs individual fields?

Reply:
If we are trying to assert that only a fixed set of fields are allowed, serializing is simpler. 


("we need both" is not a useful)

> If I compared the serialized subjects this way the tests would always fail. because the orgs wouldn't match.

this is maybe a stupid question, but _why_ are we populating the org in the stub, comparing the actual to the stub, but for org-and-only-org, testing for _in_equality?

Reply:
I would note that ToRDNSequence in go does not encode all possible fields anyway - but a fixed subset: https://pkg.go.dev/crypto/x509/pkix#Name.ToRDNSequence

So if we want to be exhaustive for all time, we can't really rely on that either.

Reply:
(could also split the difference and explicitly test everything but ExtraFields, then serialize those and compare)

Reply:
Right now, the Istio CSR we generate in the sidecar in `GenCSR` https://github.com/istio/istio/blob/master/security/pkg/pki/util/generate_csr.go#L42 has an `Org` value of `""`. So, the actual CSR we are making comparisons to has the `Org` value of `""`. `len(orgCSR.Subject.Organization) > 0 return false` would always fail. I have not thought through the implications of removing this org setting from the CSR. My initial thought would be there wouldn't be any impact, I am just cautious of an istio-proxy change because it could be significantly more impactful.

Reply:
> So, the actual CSR we are making comparisons to has the Org value of "".

weird. wonder why.

So this just boils down to `go` and `openssl` having slightly different opinons about what's allowed in `Subject.Org`? can we just gen + dump the CSR bytes with `go` then, for the tests?

Reply:
As far as I know that is the case. Yes, we could use `GenCSR` and we do for the `CreateFakeCSR` https://github.com/istio/istio/pull/51966/files#diff-9eb462ac51166aef4fdb4ca1198d23865c737c5a1a2bd8a7b4f8da9f6e1bab27R229. The problem is `GenCSR` only supports specific configuration options. We have access to set any cert options, but `GenCSR` and the subsequent methods only use a subset of those values. For example, DNSNames isn't used and I can't set CA to true.

Reply:
Perhaps an empty string slice is allowed just not configurable via the `openssl` cli. Might just be a CLI limitation? No idea.

Reply:
Hmm. Yeah really `GenCSR` should take the result of `GenCSRTemplate` as an arg, since the latter is really the only spot where the "Istio CSR opinions" live. That's a bit churny to change tho, granted.

I'd just create a `genTestCSR` in the test suite then, that uses `x509.CertificateRequest` directly, so you can set whatever you like to build "invalid" CSRs. That shouldn't be more than a few LOC in a test func, and we want to test with non-Istio CSR opinions anyway.

Reply:
Updated, let me know what you think. This adds a lot more code specific to testing but I agree it is nice to generate the CSRs programmatically.

Reply:
> This adds a lot more code specific to testing but I agree it is nice to generate the CSRs programmatically.

Yes, if that is a concern for people I would suggest that's a signal we should actually fix `GenCSR` so we can actually provide our own `GenCSRTemplate` template to it under test, and be able to properly use `GenCSR` in real impl as well as tests. 

If you write tests when you write code, you write more testable code. Or you don't, and then have to do this years later :D

LGTM and TY - the other added benefit is we don't need to check in hardcoded `openssl` blobs anymore.

Reply:
My understanding is that Subject can't be empty - or shouldn't be, because someone may use it. Note that cert chains work by using Issuer and Subject field, and the intent was to represent the actual subject just like the sub field in json. SAN literally means 'alternate' name - and it doesn't even have an equivalent in json.

K8S client certs have a well defined way to use the org ( groups) and CN ( system:serviceaccount:NAMESPACE:KSA) - same as the sub in json. But we live in the spiffe world where an opaque URL is the only thing and all other standards don't matter